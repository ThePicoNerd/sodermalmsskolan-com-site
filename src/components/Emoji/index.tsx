import React, { FunctionComponent, ReactNode } from "react";
import twemoji from "twemoji";

export interface EmojiProps {
  children: ReactNode;
}

/**
 * Twemoji-powered emoji renderer to make sure they look the same on all devices.
 *
 * @param {React.PropsWithChildren<EmojiProps>} props Emoji props.
 *
 * @returns {React.ReactElement} Rendered emoji text.
 */
const Emoji: FunctionComponent<EmojiProps> = ({
  children,
}) => (
  <>
    {React.Children.map(children, (child) => {
      if (typeof child === "string") {
        return (
          <span
            css={{
              img: {
                height: "1em",
                width: "1em",
                margin: "0 0.05em 0 0.1em",
                verticalAlign: "-0.1em",
              },
            }}
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{
              __html: twemoji.parse(child, {
                folder: "svg",
                ext: ".svg",
              }),
            }}
          />
        );
      }

      return child;
    })}
  </>
);

export default Emoji;
