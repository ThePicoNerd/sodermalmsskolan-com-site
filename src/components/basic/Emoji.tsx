import React from "react";
import styled from "styled-components";
import twemoji from "twemoji";

const EmojiWrapper = styled.span`
  img {
    height: 1em;
    width: 1em;
    margin: 0 0.05em 0 0.1em;
    vertical-align: -0.1em;
  }
`;

export const Emoji: React.FunctionComponent<{ children: React.ReactNode }> = ({
  children,
}) => {
  return (
    <>
      {React.Children.map(children, (child) => {
        if (typeof child == "string") {
          return (
            <EmojiWrapper
              dangerouslySetInnerHTML={{
                __html: twemoji.parse(child, {
                  folder: "svg",
                  ext: ".svg",
                }),
              }}
            />
          );
        }

        return child;
      })}
    </>
  );
};
