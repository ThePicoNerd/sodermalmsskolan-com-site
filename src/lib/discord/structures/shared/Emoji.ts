import { IDiscordAPIUser } from "./User";

export interface IDiscordAPIEmoji {
  id?: string;
  name?: string;
  roles?: string[];
  user?: IDiscordAPIUser;
  require_colons?: boolean;
  managed?: boolean;
  animated?: boolean;
  available?: boolean;
}
