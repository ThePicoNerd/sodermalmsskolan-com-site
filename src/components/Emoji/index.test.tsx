import { render } from "@testing-library/react";
import React from "react";
import Emoji from ".";

describe("<Emoji /> test", () => {
  it("should render standalone emojis correctly", () => {
    const { asFragment } = render(<Emoji>🐢</Emoji>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render inline emojis correctly", () => {
    const { asFragment } = render(
      // eslint-disable-next-line jsx-a11y/accessible-emoji
      <Emoji>
        Hello 🐢.
        <span>
          Do
          <span>
            you
            <span>
              even
              <span>
                nest
                {/* eslint-disable-next-line jsx-a11y/accessible-emoji */}
                <span>
                  🐢
                  <span>?</span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </Emoji>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
