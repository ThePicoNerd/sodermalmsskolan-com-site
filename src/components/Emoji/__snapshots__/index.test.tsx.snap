// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emoji test should render inline emojis correctly 1`] = `
Array [
  .emotion-0 img {
  height: 1em;
  width: 1em;
  margin: 0 0.05em 0 0.1em;
  vertical-align: -0.1em;
}

<span
    className="emotion-0"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Hello <img class=\\"emoji\\" draggable=\\"false\\" alt=\\"üê¢\\" src=\\"https://twemoji.maxcdn.com/v/13.0.1/svg/1f422.svg\\"/>.",
      }
    }
  />,
  <span>
    Do
    <span>
      you
      <span>
        even
        <span>
          nest
          <span>
            üê¢
            <span>
              ?
            </span>
          </span>
        </span>
      </span>
    </span>
  </span>,
]
`;

exports[`emoji test should render standalone emojis correctly 1`] = `
.emotion-0 img {
  height: 1em;
  width: 1em;
  margin: 0 0.05em 0 0.1em;
  vertical-align: -0.1em;
}

<span
  className="emotion-0"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<img class=\\"emoji\\" draggable=\\"false\\" alt=\\"üê¢\\" src=\\"https://twemoji.maxcdn.com/v/13.0.1/svg/1f422.svg\\"/>",
    }
  }
/>
`;
